mopidy/m3u/playlists.py

```diff
@@ -1,117 +1,117 @@
+        self._default_encoding = ext_config['default_encoding']
+        self._default_extension = ext_config['default_extension']
 
-    def get_items(self, uri):
-        playlist = self._playlists.get(uri)
-        if playlist is None:
-            return None
-        return [Ref.track(uri=t.uri, name=t.name) for t in playlist.tracks]
+    def as_list(self):
+        result = []
+        for entry in os.listdir(self._playlists_dir):
+            if not entry.endswith((b'.m3u', b'.m3u8')):
+                continue
+            elif not os.path.isfile(self._abspath(entry)):
+                continue
+            else:
+                result.append(translator.path_to_ref(entry))
+        result.sort(key=operator.attrgetter('name'))
+        return result
 
     def create(self, name):
-        playlist = self._save_m3u(Playlist(name=name))
-        self._playlists[playlist.uri] = playlist
-        logger.info('Created playlist %s', playlist.uri)
-        return playlist
+        path = translator.path_from_name(name.strip(), self._default_extension)
+        try:
+            with self._open(path, 'w'):
+                pass
+            mtime = os.path.getmtime(self._abspath(path))
+        except EnvironmentError as e:
+            log_environment_error('Error creating playlist %s' % name, e)
+        else:
+            return translator.playlist(path, [], mtime)
 
     def delete(self, uri):
-        if uri in self._playlists:
-            path = translator.playlist_uri_to_path(uri, self._playlists_dir)
-            if os.path.exists(path):
-                os.remove(path)
-            else:
-                logger.warning(
-                    'Trying to delete missing playlist file %s', path)
-            del self._playlists[uri]
-            logger.info('Deleted playlist %s', uri)
+        path = translator.uri_to_path(uri)
+        try:
+            os.remove(self._abspath(path))
+        except EnvironmentError as e:
+            log_environment_error('Error deleting playlist %s' % uri, e)
+
+    def get_items(self, uri):
+        path = translator.uri_to_path(uri)
+        try:
+            with self._open(path, 'r') as fp:
+                items = translator.load_items(fp, self._playlists_dir)
+        except EnvironmentError as e:
+            log_environment_error('Error reading playlist %s' % uri, e)
         else:
-            logger.warning('Trying to delete unknown playlist %s', uri)
+            return items
 
     def lookup(self, uri):
-        return self._playlists.get(uri)
+        path = translator.uri_to_path(uri)
+        try:
+            with self._open(path, 'r') as fp:
+                items = translator.load_items(fp, self._playlists_dir)
+            mtime = os.path.getmtime(self._abspath(path))
+        except EnvironmentError as e:
+            log_environment_error('Error reading playlist %s' % uri, e)
+        else:
+            return translator.playlist(path, items, mtime)
 
     def refresh(self):
```

Comment:
This could in theory check the mtimes of the playlists to see if they have changed, and then emit playlist changed events. Not sure if this is needed though and if it would make sense for any clients (mostly thinking about MPD ones). And if we went down that path we would probably want to consider just watching for changes instead. So perhaps just file a bug if more people think this could make some sense?